# DS2002 · Pandas Challenge

**Lab — 2026-09-18 · Fall 2026**  

---

## Lab 04 — Pandas Challenge

Four hundred generated orders. Each question builds toward a demand report you could hand a vendor.

The data is seeded, so everyone's numbers should match. That is deliberate: if your total revenue differs from your neighbor's, one of you has a bug, and the assertions at the end will tell you which.

Every answer needs the number **and** a sentence saying what it means. A cell that prints `4218.5` with no interpretation is half an answer.

In [65]:
import pandas as pd, numpy as np
rng = np.random.default_rng(4)
n = 400
df = pd.DataFrame({
    'vendor_id': rng.choice(['V-01','V-05','V-10','V-18'], n),
    'category': rng.choice(['Food','Merch','RainGear','Drink'], n, p=[.5,.2,.1,.2]),
    'qty': rng.integers(1, 4, n),
    'price': rng.choice([4.5, 6.0, 7.5, 12.0, 24.0], n),
})
df.head()

,vendor_id,category,qty,price
0,V-10,Drink,2,24.0
1,V-18,RainGear,1,12.0
2,V-18,Drink,3,4.5
3,V-10,Food,2,12.0
4,V-18,Drink,3,7.5


### Q1 — Add `revenue`, then report total revenue and total units.

*Expected: 400 rows, and revenue should land between $8,000 and $9,000.*

In [66]:
# TODO
df['revenue'] = df['qty'] * df['price'] # adds revenue to the df - calculated by the quantity & price

total_revenue = df['revenue'].sum() # total rev is the sum of the revenue variable
total_units = df['qty'].sum() # total units sums the quantity across all orders

print(df.shape[0]) # printing the number of rows - verifying 400
print(total_revenue) # printing the total revenue (between 8 and 9k)
print(total_units) # printing total units

print(f'The number of rows is calculated using the shape function. The total revenue is calculated by the quantity * price and is between 8 and 9 thousand. There are 783 total units sold across all orders')

400
8520.0
783
The number of rows is calculated using the shape function. The total revenue is calculated by the quantity * price and is between 8 and 9 thousand. There are 783 total units sold across all orders


### Q2 — Revenue by category, highest to lowest.

Include the share of total as a percentage in the same table.

In [67]:
# TODO

# Creating by_category that groups by the category's revenue total value and then sorts by highest to lowest with ascending= False
by_category = df.groupby('category')[['revenue']].sum().sort_values('revenue', ascending=False)

# Calculates the total share as a percent by taking each category's revenue and dividing it by the sum of all categories revenue and multiplying by 10 to get % value and rounding to 1 decimal place.
category_percent = (by_category['revenue'] / by_category['revenue'].sum() * 100).round(1)

# Taking by_category's revenue column and category_percent and turning them into a dataframe to present as a table
table_values = pd.DataFrame({'Revenue': by_category['revenue'], 'Percent': category_percent})

# Printing the table values
print(table_values)

print(f"First, we grouped by the category's revenue and sorted with ascending=False to make it sort from highest to lowest. Then, we determined each category's percent share by taking each category's revenue, dividing it by the total of all categories, multiplying by 100, and rounding. We then turned it into a DataFrame and printed the table.")

          Revenue  Percent
category                  
Food       4293.0     50.4
Merch      1771.5     20.8
Drink      1554.0     18.2
RainGear    901.5     10.6
First, we grouped by the category's revenue and sorted with ascending=False to make it sort from highest to lowest. Then, we determined each category's percent share by taking each category's revenue, dividing it by the total of all categories, multiplying by 100, and rounding. We then turned it into a DataFrame and printed the table.


### Q3 — Which vendor has the highest *average* order revenue?

Report the average alongside the order count for each vendor. A high average on twelve orders is a different claim from a high average on two hundred.

In [68]:
# TODO
# Creating the by_vendor variable to be group the dataframe by the vendor id and revenue. We then aggregate the data by the mean & count and sort it based on the mean's from highest to lowest.
by_vendor = df.groupby('vendor_id')['revenue'].agg(['mean', 'count']).sort_values('mean', ascending = False)

# Created by_vendor.columns to provide names for the dataframe
by_vendor.columns = ['average_revenue', 'order_count']

# Prints out the dataframe
print(by_vendor)

print('Uses the by_vendor variable to consider the vendor_id and revenue while aggregating by the mean and count, and ultimately sorting by the mean from highest to low, and then adding the column names for the dataframe, and then printing the dataframe. Vendor 1 has the highest mean order revenue ($22.60) and it is based on 94 orders; however, vendor 18 has 108 orders and a lower average.')

           average_revenue  order_count
vendor_id                              
V-01             22.595745           94
V-18             21.750000          108
V-05             20.580645           93
V-10             20.314286          105
Uses the by_vendor variable to consider the vendor_id and revenue while aggregating by the mean and count, and ultimately sorting by the mean from highest to low, and then adding the column names for the dataframe, and then printing the dataframe. Vendor 1 has the highest mean order revenue ($22.60) and it is based on 94 orders; however, vendor 18 has 108 orders and a lower average.


### Q4 — What share of revenue comes from Merch?

Print it as a percentage rounded to one decimal.

In [69]:
# TODO

# Creating the merch_share value by divding the merch value by the all the categories and multiplying it by 100 to get the % and then rounding it to 1 decimal place.
merch_share = (by_category.loc['Merch', 'revenue'] / by_category['revenue'].sum() * 100).round(1)

# Printing the merch share
print(merch_share)

print("Calculating the merch_share by taking merch's revenue from by_category and dividing it by the total of all categories, and then multiplying by 100 to get a percentage value, and rounding to the first decimal place. The merch accounts for 20.8% of total revenue.")

20.8
Calculating the merch_share by taking merch's revenue from by_category and dividing it by the total of all categories, and then multiplying by 100 to get a percentage value, and rounding to the first decimal place. The merch accounts for 20.8% of total revenue.


### Q5 — Join in the vendor names.

The frame only has `vendor_id`. Merge the lookup below so your report is readable.

**Requirements:** left join, `validate='many_to_one'`, and prove the row count and revenue total did not change. One vendor id in the orders is not in this lookup — find it, and decide what to do about it.

In [70]:
vendor_names = pd.DataFrame({
    'vendor_id': ['V-01', 'V-05', 'V-10'],
    'vendor_name': ['Hoos Burgers', 'Rotunda Tacos', 'Cav Merch North'],
})

# TODO: merge, validate, and report the unmatched vendor

# Using a left_join where we use the df and consider the vendor names. Join on the vendor_id with a left join and use validate on "many_to_one"
left_join = pd.merge(df, vendor_names, on='vendor_id', how='left', validate='many_to_one')

# Checking the row count from the left_join and df
print(len(left_join)== len(df))

# Checking the revenue with the sum of the left_join and the df
print(left_join['revenue'].sum() == df['revenue'].sum())

# Determine the non-matching name - checking the left_join vendor name and seeing if there is a null value and determining if their vendor_id is unique
missing = left_join[left_join['vendor_name'].isna()]['vendor_id'].unique()

# Printing the missing/ non-matching name
print(missing)

# Solution - assigning the vendor name in the left join as unknown vendor
left_join['vendor_name'] = left_join['vendor_name'].fillna('Unknown Vendor (V-18)')

True
True
['V-18']


**The unmatched vendor, and what I did about it:** _..._

Vendor 18 or V-18 does not have a matching name in the table. To remedy the missing value, I created a placeholder (similar to the studio check point) value rather than dropping. I wanted to avoid dropping it because it would reduce the row count and revenue.



### Q6 — A pivot table: vendors down the side, categories across the top, revenue in the cells.

Add row and column totals so it reads as a report rather than a grid of numbers.

In [71]:
# TODO

# Creating a pivot table with pivot_table function - taking in the df values, where the values, index, columns, aggfunc, margins, and margins_name are assigned values
pivot = pd.pivot_table(df,
                       values='revenue',
                       index='vendor_id',
                       columns='category',
                       aggfunc='sum',
                       margins=True,
                       margins_name='Total')

# Printing the pivot table
print(pivot)

print('Added the rows and columns using the pivot_table function. Vendor V-18 earned the most revenue; however, it did not have the highest average order (found in Q3)')

category    Drink    Food   Merch  RainGear   Total
vendor_id                                          
V-01        171.0  1338.0   373.5     241.5  2124.0
V-05        298.5   882.0   489.0     244.5  1914.0
V-10        502.5  1054.5   400.5     175.5  2133.0
V-18        582.0  1018.5   508.5     240.0  2349.0
Total      1554.0  4293.0  1771.5     901.5  8520.0
Added the rows and columns using the pivot_table function. Vendor V-18 earned the most revenue; however, it did not have the highest average order (found in Q3)


### Q7 — Validate your work

**TODO:** uncomment and make these pass. Assign your results to the named variables as you go.

In [72]:
assert len(df) == 400
assert 8000 < df['revenue'].sum() < 9000, df['revenue'].sum()
assert abs(by_category['revenue'].sum() - df['revenue'].sum()) < 0.01
assert len(left_join) == len(df), 'the vendor merge changed the row count'
print('checks passed.')

checks passed.


### Write-up

**a)** What would you tell these vendors to do differently next game? One paragraph, with at least two numbers from your report in it.

**b)** Which of your seven answers is the least trustworthy, and why? Point at a specific weakness — a small group size, an unmatched vendor, a category that is really two things.

**Question A**

Looking at the revenues, the vendors should reconsider where their efforts are being placed. The food category makes up 50.4% of their revenue; similarly, they should focus on their merch as well. However, they should possibly remove shelf space or buy less RainGear, considering it only makes up 10.6% of their revenue. In addition, when looking at a vendor level, V-05 is relatively underperforming, as it has the lowest revenue of all four vendors, making ~$210 less than V-01, the vendor with the third-highest revenue. That vendor should possibly run promotions or change their tactics to bring in more revenue.

**Question B**

In question three, we calculated the average order revenue by vendor. While averaging is typically a safe approach, there were many orders within the dataset (93 to 108), in which one of the orders could have an outlier, thus impacting the average. The average is typically more sensitive to outliers than a total, since an unusually large order could skew an average upward:, thus disproportionately representing the vendor's revenue; however, the same order does not shift a sum in the same way across many orders, in comparison to the average. That said, the vendors' average values range from 20.31 to 22.60 dollars, thus showing the discrepancies between vendors, possibly impacted by variation within the orders, therefore making it one of the weaker calculations within the notebook.